update docs on schema (#256)
catalystneuro/neuroconv: docs/developer_guide/schemas.rst

.. _source_schema:

Defining the form of source with schemas
----------------------------------------

The ``.__init__()`` of each ``DataInterface`` has a unique call signature, requiring the source data to be input in a

> specific way. Some require a ``file_path``, some a ``directory_path``, and some require additional arguments that are

Reviewer: ```suggestion
specific way. Some require a ``file_path``, some a ``folder_path``, and some require additional arguments that are
```

necessary to read the source data properly. The same is true for the ``.run_conversion()`` method, where some classes
have unique additional optional arguments. To communicate these requirements and options with an ``NWBConverter`` and
with interfacing software, the expected form of the input is defined using the JSON Schema language.
``DataInterface`` classes define the method
:py:func:`~neuroconv.datainterfaces.basedatainterface.BaseDataInterface.get_source_schema()`, which is
responsible for returning the source schema as a JSON schema dictionary. By default, these dictionaries are
automatically derived from the call signatures of a function, and express similar information to a type-hinted call

> signature, but can also be used to define and validate inputs that are arbitrarily deep nested dictionaries. For

Reviewer: ```suggestion
signature, but can also be used to define and validate inputs that are nested dictionaries of arbitrary depth. For
```
'arbitrarily deep nested dictionaries' took a couple of re-reads to understand

    {
      "required": [
        "file_path"
      ],
      "properties": {
        "file_path": {
          "format": "file",
          "type": "string",
          "description": "Path to SpikeGLX file."
        },
        "stub_test": {
          "type": "boolean",
          "default": false
        },
        "spikeextractors_backend": {
          "type": "boolean",
          "default": false
        },
        "verbose": {
          "type": "boolean",
          "default": true
        }
      },
      "type": "object",
      "additionalProperties": false
    }

``PhySortingInterface.get_source_schema()`` returns a similar schema:

.. code-block:: json

    {
      "required": [
        "folder_path"
      ],
      "properties": {
        "folder_path": {
          "format": "directory",
          "type": "string"
        },
        "exclude_cluster_groups": {
          "type": "array"
        },
        "verbose": {
          "type": "boolean",
          "default": true
        },
        "spikeextractors_backend": {
          "type": "boolean",
          "default": false
        }
      },
      "type": "object",
      "additionalProperties": false
    }

> A ``ExampleNWBConverter`` that combines these ``DataInterface``s will combine the JSON Schema using the keys of the

Reviewer: ```suggestion
An ``ExampleNWBConverter`` that combines these ``DataInterface``s will combine the JSON Schema using the keys of the
```
Matter of opinion, but I imaging reading the Python class as if it were a part of the sentence, like 'An example NWBConverter...'

> A ``ExampleNWBConverter`` that combines these ``DataInterface``s will combine the JSON Schema using the keys of the

Reviewer: ```suggestion
A ``ExampleNWBConverter`` that combines these ``DataInterface`` s will combine the JSON Schema using the keys of the
```
Something is causing the beginning '``' of the `DataInterface` to not end until after the '``' of `data_interface_classes` - wondering if this is it

Reviewer: ![image](https://user-images.githubusercontent.com/51133164/208479170-93619eba-cd13-46b4-9cfc-ecbe515444fc.png)

Reviewer: Huh, so I guess adding the space will 'fix it' but then there's an awkward amount of space between the end of the code and the `s` character. Any idea how to fix this? Maybe just a rephrase to avoid the issue entirely: like 'An `ExampleNWBConverter that combines multiple `DataInterface` objects...'

            },
            "verbose":{
              "type":"boolean",
              "default":true
            }
          },
          "type":"object",
          "additionalProperties":false
        },
        "PhySorting":{
          "required":[
            "folder_path"
          ],
          "properties":{
            "folder_path":{
              "format":"directory",
              "type":"string"
            },
            "exclude_cluster_groups":{
              "type":"array"
            },
            "verbose":{
              "type":"boolean",
              "default":true
            },
            "spikeextractors_backend":{
              "type":"boolean",
              "default":false
            }
          },
          "type":"object",
          "additionalProperties":false
        }
      },
      "additionalProperties":false
    }

Conversion schemas options work similarly to source schemas.

.. _metadata_schema:

Metadata Schema
---------------

Similar to input data, each ``DataInterface`` produces its own metadata schema reflecting
the specificities of the dataset it interfaces with. The ``DataInterface``-specific metadata schema can be obtained

> via method ``.get_metadata_schema()``. Unlike ``.get_source_schema``, the ``DataInterface`` needs to be initialized

Reviewer: ```suggestion
via the ``.get_metadata_schema()`` method . Unlike ``.get_source_schema``, the ``DataInterface`` needs to be initialized
```

> via method ``.get_metadata_schema()``. Unlike ``.get_source_schema``, the ``DataInterface`` needs to be initialized

Reviewer: ```suggestion
via method ``.get_metadata_schema()``. Unlike ``.get_source_schema()``, the ``DataInterface`` needs to be initialized
```

For consistency with the previous line

Reviewer: ![image](https://user-images.githubusercontent.com/51133164/208479418-ef5f95eb-617e-47d5-8048-610c5fa6e0b1.png)

                },
                "control": {
                  "description": "Numerical labels that apply to each element in data",
                  "type": "array"
                },
                "control_description": {
                  "description": "Description of each control value",
                  "type": "array"
                },
                "offset": {
                  "description": "Scalar to add to each element in the data scaled by 'conversion' to finish converting it to the specified unit.",
                  "type": "number",
                  "default": 0
                }
              },
              "type": "object",
              "additionalProperties": false,
              "tag": "pynwb.ecephys.ElectricalSeries"
            }
          },
          "type": "object",
          "additionalProperties": false,
          "tag": "Ecephys"
        }
      },
      "type": "object",
      "additionalProperties": false,
      "$schema": "http://json-schema.org/draft-07/schema#",
      "$id": "metadata.schema.json",
      "title": "Metadata",
      "description": "Schema for the metadata",
      "version": "0.1.0"
    }

> Like with the source schema, metadata schema are combined automatically when combined in an ``NWBConverter``, which

Reviewer: ```suggestion
Like with the source schema, metadata schema are combined across ``DataInterface`` s automatically in an ``NWBConverter``, which
```

> can also be accessed using ``.get_metadata_schema``, though the underlying schemas are merged directly with each other

Reviewer: ```suggestion
can also be accessed using ``.get_metadata_schema()``, though the underlying schemas are merged directly with each other
```
For consistency with above usage - or the above usage can also just remove the `()`, either way

> instead of being combined on a key basis.

Reviewer: Hmm... This point isn't super clear, but if your intent isn't to get too deep into the distinction, at a minimum I'd include something like 'For more information on how these nested dictionaries automatically merge, refer to https://neuroconv--256.org.readthedocs.build/en/256/api/utils.html#neuroconv.utils.dict.dict_deep_update' (with a better sphinx link to that API doc, of course)

.. _source_schema:

Defining the form of source with schemas
----------------------------------------

The ``.__init__()`` of each ``DataInterface`` has a unique call signature, requiring the source data to be input in a
specific way. Some require a ``file_path``, some a ``folder_path``, and some require additional arguments that are
necessary to read the source data properly. The same is true for the ``.run_conversion()`` method, where some classes
have unique additional optional arguments. To communicate these requirements and options with an ``NWBConverter`` and
with interfacing software, the expected form of the input is defined using the JSON Schema language.
``DataInterface`` classes define the method
:py:func:`~neuroconv.datainterfaces.basedatainterface.BaseDataInterface.get_source_schema()`, which is
responsible for returning the source schema as a JSON schema dictionary. By default, these dictionaries are

> automatically derived from the call signatures of a function, and express similar information to a type-hinted call
> signature, but can also be used to define and validate inputs that are nested dictionaries of arbitrary depth. For

Author: ```suggestion
automatically derived from the call signatures and type hints of a function, and express similar information, 
but can also be used to define and validate inputs that are nested dictionaries of arbitrary depth. For
```

example, :py:func:`SpikeGLXRecordingInterface.__init__()` has the call signature

.. code-block:: python

    def __init__(
        self,

>         file_path: str,

Author: ```suggestion
        file_path: FilePathType,
```

        stub_test: bool = False,
        spikeextractors_backend: bool = False,
        verbose: bool = True,
    )

Calling
:py:func:`~neuroconv.datainterfaces.ecephys.spikeglx.spikeglxdatainterface.SpikeGLXRecordingInterface.get_source_schema()`

> gives the following output, which is derived from the call signature and expresses identical constraints.

Author: ```suggestion
gives the following output, which is derived from the call signature and expresses identical information.
```

    {
      "required": [
        "file_path"
      ],
      "properties": {
        "file_path": {
          "format": "file",
          "type": "string",
          "description": "Path to SpikeGLX file."
        },
        "stub_test": {
          "type": "boolean",
          "default": false
        },
        "spikeextractors_backend": {
          "type": "boolean",
          "default": false
        },
        "verbose": {
          "type": "boolean",
          "default": true
        }
      },
      "type": "object",
      "additionalProperties": false
    }

``PhySortingInterface.get_source_schema()`` returns a similar schema:

.. code-block:: json

    {
      "required": [
        "folder_path"
      ],
      "properties": {
        "folder_path": {
          "format": "directory",
          "type": "string"
        },
        "exclude_cluster_groups": {
          "type": "array"
        },
        "verbose": {
          "type": "boolean",
          "default": true
        },
        "spikeextractors_backend": {
          "type": "boolean",
          "default": false
        }
      },
      "type": "object",
      "additionalProperties": false
    }

> An ``ExampleNWBConverter`` that combines these ``DataInterface``s will combine the JSON Schema using the keys of the

Author: ```suggestion
An ``ExampleNWBConverter`` that combines these ``DataInterface`` classes will combine the JSON Schema using the keys of the
```

            },
            "verbose":{
              "type":"boolean",
              "default":true
            }
          },
          "type":"object",
          "additionalProperties":false
        },
        "PhySorting":{
          "required":[
            "folder_path"
          ],
          "properties":{
            "folder_path":{
              "format":"directory",
              "type":"string"
            },
            "exclude_cluster_groups":{
              "type":"array"
            },
            "verbose":{
              "type":"boolean",
              "default":true
            },
            "spikeextractors_backend":{
              "type":"boolean",
              "default":false
            }
          },
          "type":"object",
          "additionalProperties":false
        }
      },
      "additionalProperties":false
    }

Conversion schemas options work similarly to source schemas.

.. _metadata_schema:

Metadata Schema
---------------

Similar to input data, each ``DataInterface`` produces its own metadata schema reflecting
the specificities of the dataset it interfaces with. The ``DataInterface``-specific metadata schema can be obtained

> via the ``.get_metadata_schema()`` method . Unlike ``.get_source_schema()``, the ``DataInterface`` needs to be

Author: ```suggestion
via the ``.get_metadata_schema()`` method. Unlike ``.get_source_schema()``, the ``DataInterface`` needs to be
```
